### ISSUE TITLE
Primary Key Generator

### ISSUE THREAD

**Issue by splitbrain** [tflori/orm#47]
I'm currently refactoring a table from an auto increment integer PK to a UUID based one. Since the entity is be passed around in the code before it is actually saved, I have some guardians at certain places to make sure to not act on a non-saved entity. So far I did this with a simple `if($entity->id)` check. I would like to continue to do this, so I should only generate the primary key UUID at the latest possible moment before saving.

I do so in the prePersist() method:

```php
    public function prePersist()
    {
        if(!$this->id) {
            $this->id = genreateUUID();
        }
    }
```

Now, since my column isn't an auto-incremented integer anymore, I also set

```php
protected static $autoIncrement = false;
```

This will throw an `IncompletePrimaryKey` exception in the save method *before* calling `prePersist()`. For now I simply did not declare the `$autoIncrement = false` and it's working. But this seems a bit dirty to me.

So I guess the question is: what would be the clean way to create a Primary Key generator that is only called right before insert?

**Comment by splitbrain** [comment]
Actually it did not work. While the correct UUID primary key was saved, the Entity returned by the `save()` method contained an integer primary key (probably the rowid - I'm using sqlite).

So I now *do* set `$autoIncrement = false` and simply overwrite the save method. A bit less dirty but maybe not as elegant as it should be:

```php
    public function save()
    {
        if(!$this->id) {
            $this->id = generateUUID();
        }
        return parent::save();
    }
```

**Comment by tflori** [comment]
Valid point didn't thought about it. So the increment check should be
executed after prePersist or the check tries a generateId method or
similar...

Andreas Gohr <notifications@github.com> schrieb am Mo., 3. Sep. 2018, 15:51:

**Comment by tflori** [comment]
https://github.com/tflori/orm/blob/master/src/Entity.php#L578

maybe here just add an `elseif (is_callable([$this, 'generatePrimaryKey']))`

**Comment by tflori** [comment]
@splitbrain solved with version 1.4

you can now use the interface and overload `protected function generatePrimaryKey()`:

```php
class Article extends Entity implements Entity\GeneratesPrimaryKeys
{
  protected static $autoIncrement = false;
  protected function generatePrimaryKey()
  {
    $this->id = generateUUID();
  }
}
```

The method will only be called when the primary key is incomplete - so no need to check if id is empty.

Thanks for your contribution! Feel free to reopen if something is wrong with the implementation.

**Comment by splitbrain** [comment]
awesome! thank you.
